### PR TITLE
noice: migrate to mkNeovimPlugin

### PR DESCRIPTION
**Pull request by khaneliman**
~Got the main migration done. But, there are tons of options without descriptions or not appropriately formatted descriptions, atm.~

### CODE REVIEW

**Comment by MattSturgeon** [comment]
> But, there are tons of options without descriptions or not appropriately formatted descriptions, atm.

I haven't looked, but if these options are not too important and do not have any type coercion, perhaps we could consider dropping them and relying on the freeform type.

Especially given they do not even have documentation...

File: plugins/by-name/noice/default.nix [hunk 1]
```diff
@@ -1,35 +1,103 @@
 {
   lib,
-  helpers,
-  config,
-  pkgs,
   ...
 }:
-# TODO: This uses a lot of types.anything because noice.nvim types are quite complex.
-# It should be possible to map them to nix, but they would not map really well through
-# toLuaObject, we would maybe need some ad-hoc pre-processing functions.
```
Comment (MattSturgeon) [inline plugins/by-name/noice/default.nix:10]: Is this comment still relevant? Or was it removed intentionally?

File: plugins/by-name/noice/default.nix [hunk 2]
```diff
@@ -1,35 +1,103 @@
...
+  originalName = "noice.nvim";
+  package = "noice-nvim";
+
+  maintainers = [ lib.maintainers.khaneliman ];
+
+  description = ''
+    `noice.nvim`, an experimental Neovim UI.
+
+    > [!NOTE] If treesitter is enabled you need the following parsers:
+    >  vim, regex, lua, bash, markdown, markdown_inline
```
Comment (MattSturgeon) [inline plugins/by-name/noice/default.nix:20]:
```suggestion
    > [!NOTE]
    > If treesitter is enabled you need the following parsers:
    > vim, regex, lua, bash, markdown, markdown_inline
```

I'm unsure if your initial syntax is valid, but elsewhere we have the admonition syntax on a separate line from the actual text body

File: plugins/by-name/noice/default.nix [hunk 3]
```diff
@@ -1,35 +1,103 @@
...
 
-    package = lib.mkPackageOption pkgs "noice" {
-      default = [
-        "vimPlugins"
-        "noice-nvim"
+  # TODO: added 2024-10-27 remove after 24.11
+  deprecateExtraOptions = true;
+  optionsRenamedToSettings =
+    let
+      mkOptionPaths = paths: map (p: lib.splitString "." p) paths;
```
Comment (MattSturgeon) [inline plugins/by-name/noice/default.nix:27]:
```suggestion
      mkOptionPaths = map (lib.splitString ".");
```

No need to specify the arg if it is passed as the final arg anyway. We can just use a partially applied function.

Same for the mapping function.

File: plugins/by-name/noice/default.nix [hunk 4]
```diff
@@ -1,35 +1,103 @@
...
+        "presets"
+        "throttle"
+        "views"
+        "routes"
+        "status"
+        "format"
       ];
-    };
+    in
+    mkOptionPaths rawPaths;
```
Comment (MattSturgeon) [inline plugins/by-name/noice/default.nix:81]: Stylistic: you _could_ inline `rawPaths` here if you wanted.

**Review by MattSturgeon — COMMENTED** [review]
Just some initial thoughts, as I don't have time to properly review right now.

File: plugins/by-name/noice/default.nix (hunk 2 above)
Comment (khaneliman) [reply plugins/by-name/noice/default.nix:20]: Doh, you're right 
image

File: plugins/by-name/noice/default.nix (hunk 1 above)
Comment (khaneliman) [reply plugins/by-name/noice/default.nix:10]: I guess I wasn't sure of its relevance anymore if we go down the route of allowing freeform settings and only type the ones that seem obvious. I'm double checking types.

File: plugins/by-name/noice/default.nix [hunk 5]
```diff
@@ -165,78 +239,93 @@ with lib;
...
 
-        throttle = helpers.defaultNullOpts.mkNum (literalExpression "1000 / 30") "frequency to update lsp progress message";
+        format_done =
+          defaultNullOpts.mkNullableWithRaw (with types; either str anything) "lsp_progress_done"
+            ''
+              Lsp Progress is formatted using the builtins for lsp_progress.
+            '';
 
-        view = helpers.defaultNullOpts.mkStr "mini" "";
+        throttle = defaultNullOpts.mkNum (lib.literalExpression "1000 / 30") "Frequency to update lsp progress message.";
```
Comment (MattSturgeon) [inline plugins/by-name/noice/default.nix:274]: I think this is blamed on me, but I wonder why we're saying `1000 / 30` instead of the simpler `100 / 3`?

**Review by MattSturgeon — APPROVED** [review]
I don't see any issues after a quick read through the diff.

Thanks again :rocket:

File: plugins/by-name/noice/default.nix (hunk 5 above)
Comment (khaneliman) [reply plugins/by-name/noice/default.nix:274]: I thought I saw this in the upstream default so maybe just respecting what they had?

Comment (MattSturgeon) [reply plugins/by-name/noice/default.nix:274]: Yeah, I guess so. Just seems strange to have the extra `0` given `(100 * N) / (3 * N) = 100 / 3 ≈ 33.33`

**Comment by khaneliman** [comment]
@mergify queue

File: plugins/by-name/noice/default.nix (hunk 1 above)
Comment (traxys) [reply plugins/by-name/noice/default.nix:10]: The context is that noice types are horrible to map to nix (mixed list/attrs, recursive,...)

With freeform options this is much less of an issue

Comment (MattSturgeon) [reply plugins/by-name/noice/default.nix:10]: > With freeform options this is much less of an issue

Unless we inadvertently impose type-restrictions by declaring settings options 🙃

**Pull request merged**